### DISCUSSION TITLE
Accessing GH context in actions

### DISCUSSION THREAD

**Discussion by jnus** [community/community#26326] [Actions]
<p>Is there anyway to access the GH context from within an action, without having to specify e.g. env og secrets as inputs for the action? Right now I serialize e.g. secrets as a parameter for the action and deserialize within the action.</p>
<p>Thoughts?</p>

**Comment by jnus** [comment]
<p>Can anyone confirm that accessing GH context is not possible within an action, without specifying it as a parameter input?</p>

**Comment by rcowsill** [comment, marked as answer]
<p>I don’t think it’s currently possible to access the context directly.</p>
<p>Consider <a href="https://github.com/actions/github-script" rel="noopener nofollow ugc">github-script</a>, which provides a partial context containing some of the items from the workflow github context. That partial context is actually just rehydrated from the environment. If it were possible to get the real context directly it would make sense for github-script to do that. The fact it doesn’t suggests it’s not possible.</p>
<p>One workaround is to add inputs to your action, but set the default using the context, eg:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/actions/github-script/blob/00e1b58cd9e040d944127e75b3d9df586e4b0479/action.yml#L13">
  <header class="source">

      <a href="https://github.com/actions/github-script/blob/00e1b58cd9e040d944127e75b3d9df586e4b0479/action.yml#L13" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/actions/github-script/blob/00e1b58cd9e040d944127e75b3d9df586e4b0479/action.yml#L13" target="_blank" rel="noopener nofollow ugc">actions/github-script/blob/00e1b58cd9e040d944127e75b3d9df586e4b0479/action.yml#L13</a></h4>



  <pre class="onebox">    <code class="lang-yml">
      <ol class="start lines" start="3" style="counter-reset: li-counter 2 ;">
          <li>description: Run simple scripts using the GitHub client</li>
          <li>branding:</li>
          <li>  color: blue</li>
          <li>  icon: code</li>
          <li>inputs:</li>
          <li>  script:</li>
          <li>    description: The script to run</li>
          <li>    required: true</li>
          <li>  github-token:</li>
          <li>    description: The GitHub token used to create an authenticated client</li>
          <li class="selected">    default: ${{ github.token }}</li>
          <li>    required: false</li>
          <li>  debug:</li>
          <li>    description: Whether to tell the GitHub client to log details of its requests</li>
          <li>    default: false</li>
          <li>  user-agent:</li>
          <li>    description: An optional user-agent string</li>
          <li>    default: actions/github-script</li>
          <li>  previews:</li>
          <li>    description: A comma-separated list of API previews to accept</li>
          <li>  result-encoding:</li>
      </ol>
    </code>
  </pre>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>That’s ok if it makes sense to give users of the action the option to override the default value.</p>

**Comment by jnus** [comment]
<p>That’s a pretty good idea and letting the user option out by overriding the inputs makes perfectly sense. In my case I need access to both the github and secrets context for setting up variables based on the environment context in the <a href="https://github.com/marketplace/actions/json-variables" rel="noopener nofollow ugc">json-variables</a> action. The default assignment makes usage syntax far more clean, while still being able to overwrite.<br>
Thanks!</p>

**Comment by kwin** [comment]
<p>But it is explicitly stated in <a href="https://docs.github.com/en/actions/security-guides/automatic-token-authentication#using-the-github_token-in-a-workflow" class="inline-onebox" rel="noopener nofollow ugc">Automatic token authentication - GitHub Docs</a></p>
<blockquote>
<p><strong>Important:</strong> An action can access the <code>GITHUB_TOKEN</code> through the <code>github.token</code> context even if the workflow does not explicitly pass the <code>GITHUB_TOKEN</code> to the action.</p>
</blockquote>
<p>Is this just plain wrong?</p>

**Comment by josecelano** [comment]
<p>I was wondering exactly the same thing. I think the way to get the GITHUB_TOKEN using the context is using a default input value. You can see an example in the GitHub <a href="https://github.com/actions/checkout" rel="noopener nofollow ugc">actions/checkout@v2</a> action:</p>
<aside class="onebox githubblob" data-onebox-src="https://github.com/actions/checkout/blob/2541b1294d2704b0964813337f33b291d3f8596b/action.yml#L24">
  <header class="source">

      <a href="https://github.com/actions/checkout/blob/2541b1294d2704b0964813337f33b291d3f8596b/action.yml#L24" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/actions/checkout/blob/2541b1294d2704b0964813337f33b291d3f8596b/action.yml#L24" target="_blank" rel="noopener nofollow ugc">actions/checkout/blob/2541b1294d2704b0964813337f33b291d3f8596b/action.yml#L24</a></h4>



    <pre class="onebox">      <code class="lang-yml">
        <ol class="start lines" start="14" style="counter-reset: li-counter 13 ;">
            <li>    Personal access token (PAT) used to fetch the repository. The PAT is configured
</li>
            <li>    with the local git config, which enables your scripts to run authenticated git
</li>
            <li>    commands. The post-job step removes the PAT.
</li>
            <li>
</li>
            <li>
</li>
            <li>    We recommend using a service account with the least permissions necessary.
</li>
            <li>    Also when generating a new PAT, select the least scopes necessary.
</li>
            <li>
</li>
            <li>
</li>
            <li>    [Learn more about creating and using encrypted secrets](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets)
</li>
            <li class="selected">  default: ${{ github.token }}
</li>
            <li>ssh-key:
</li>
            <li>  description: &gt;
</li>
            <li>    SSH key used to fetch the repository. The SSH key is configured with the local
</li>
            <li>    git config, which enables your scripts to run authenticated git commands.
</li>
            <li>    The post-job step removes the SSH key.
</li>
            <li>
</li>
            <li>
</li>
            <li>    We recommend using a service account with the least permissions necessary.
</li>
            <li>
</li>
            <li>
</li>
        </ol>
      </code>
    </pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>As far as I know there are some ways to pass values to the action:</p>
<ul>
<li>Inputs in the step (<code>with</code>): <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepswith" class="inline-onebox" rel="noopener nofollow ugc">Workflow syntax for GitHub Actions - GitHub Docs</a>
</li>
<li>Arguments in the step (<code>args</code>): <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepswithargs" class="inline-onebox" rel="noopener nofollow ugc">Workflow syntax for GitHub Actions - GitHub Docs</a>
</li>
<li>Environments variables in the step (<code>env</code>): <a href="https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsenv" class="inline-onebox" rel="noopener nofollow ugc">Workflow syntax for GitHub Actions - GitHub Docs</a>
</li>
</ul>
<p>And:</p>
<ul>
<li>Default input value in the action definition yml file(<code>default</code>): <a href="https://github.com/actions/checkout/blob/2541b1294d2704b0964813337f33b291d3f8596b/action.yml#L24" class="inline-onebox" rel="noopener nofollow ugc">checkout/action.yml at 2541b1294d2704b0964813337f33b291d3f8596b · actions/checkout · GitHub</a>
</li>
</ul>
<p>I have not found any information in the documentation regarding using contexts in the action definition file. It is not mentioned in the action.yml syntax: <a href="https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#inputsinput_iddefault" class="inline-onebox" rel="noopener nofollow ugc">Metadata syntax for GitHub Actions - GitHub Docs</a></p>
